Add distribution constraint options to pip download

patricklaw opened the pull request:
This is a continuation of https://github.com/pypa/pip/pull/2965.  Right now this branch also includes the changes in https://github.com/pypa/pip/pull/3085, pending 3085 being merged.  This PR is to trigger Travis and to prepare for a quick followup once 3085 is merged.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/pypa/pip/3092)
<!-- Reviewable:end -->

patricklaw commented:
Working on updating this and adding some more thorough tests now

jorgebastida commented:
Hello @patricklaw I'm really looking forward to ``--platform`` to arrive to ``pip``. There is anything missing in this PR I could help with in order to move it forward? Thanks!

patricklaw commented:
Hi @jorgebastida, this ended up falling to the backburner as I got caught up in other things.  I'll try to refresh my memory on it today and report back.  Last I remember there were still some issues with plumbing the pep425 tags down.  Nothing seemed particularly monumental, just needs a little tweaking and some more tests.

patricklaw commented:
Could someone clarify what the issue is with this shard? https://travis-ci.org/pypa/pip/jobs/125431745

I wasn't particularly comfortable checking in binaries, but it seemed to be the standard practice for the test fixture.  If it's preferred, I can instead programmatically generate these wheels in a temp directory using a fixture.

patricklaw commented:
Okay, based on the failing pypy shard it seems like I should just generalize this patch to also deal with interpreter implementation and ABI.  Working on that now.

patricklaw commented:
Added those options and figured out the packaging issue (silly mistake; accidentally checked in some debugging output).  If this CI run is green then this PR is ready for review.

patricklaw commented:
It looks like a couple of shards are failing due to unrelated timeouts.  Hopefully this next run is green.  This is finished from my end and ready to be reviewed/merged.  Please let me know if there's anything else that needs to be done.

xavfernandez commented on pip/commands/download.py:
```diff
@@ -63,6 +63,67 @@ def __init__(self, *args, **kw):
+        )
+
+        cmd_opts.add_option(
+            '--manylinux1',
```
Do we really need this special case flag ?
`pip download numpy --platform=manylinux1_x86_64` could maybe do the job ?

patricklaw replied:
Agreed, with the caveat that when the user is specifying a platform, `manylinux1` should definitely be considered false, since it could otherwise leak into the tags in addition to what the user specified.

xavfernandez commented on pip/pep425tags.py:
```diff
@@ -264,12 +264,23 @@ def _supports_arch(major, minor, arch):
+        tags for, or None. If None, use the local interpreter impl.
+    :param abi: specify the exact abi you want valid
+        tags for, or None. If None, use the local interpreter abi.
+    :param manylinux1: Whether or not to add manylinux1 as a comatible
```
`comatible` -> `compatible`

xavfernandez commented on pip/commands/download.py:
```diff
@@ -63,6 +63,67 @@ def __init__(self, *args, **kw):
+        )
+
+        cmd_opts.add_option(
+            '--interpreter-version',
```
I'm not sur `interpreter-version` is the correct arg.
For example, latest PyPy interpreter version is 5.1.
It should maybe be `--python-version` ?

patricklaw replied:
Agreed.

patricklaw commented:
@xavfernandez Done!

xavfernandez commented:
This needs more documentation I think with a big warning on its limitation:
if we are using `pip dowload --python-version=27 foo` from a Python 3.5 and `foo` comes through a `.tar.gz` the `setup.py egg_info` will be executed as Python 3.5 and the downloaded dependencies might be the ones for 3.5 and not those for 2.7.

patricklaw commented:
@xavfernandez Good catch.  I changed it to enforce `--only-binary=:all:` (and disallow any `--no-binary=foo` exceptions), added a test for the enforcement, and noted this requirement in the docs.

patricklaw commented:
Travis shard failure looks spurious.

patricklaw commented:
Anything in particular blocking this being merged?

patricklaw commented:
Rebased and squashed to fix merge conflicts.

patricklaw commented:
Force pushed to resolve another merge conflict

dstufft commented:
Accidentally closed this, reopening. Sorry!

BrownTruck commented:
Hello!

As part of an effort to ease the contribution process and adopt a more standard workflow pip has switched to doing development on the ``master`` branch. However, this Pull Request was made against the ``develop`` branch so it will need to be resubmitted against ``master``. Unfortunately, this pull request does not cleanly merge against the current ``master`` branch.

If you do nothing, this Pull Request will be automatically closed by @BrownTruck since it cannot be merged.

If this pull request is still valid, please rebase it against ``master`` (or merge ``master`` into it) and resubmit it against the ``master`` branch, closing and referencing the original Pull Request.

If you choose to rebase/merge and resubmit this Pull Request, here is an example message that you can copy and paste:

```
This is a continuation of #2965.  Right now this branch also includes the changes in #3085, pending 3085 being merged.  This PR is to trigger Travis and to prepare for a quick followup once 3085 is merged.

---

*This was migrated from pypa/pip#3092 to reparent it to the ``master`` branch. Please see original pull request for any previous discussion.*
```

BrownTruck commented:
This Pull Request was closed because it cannot be automatically reparented to the ``master`` branch and it appears to have bit rotted.

Please feel free to re-open it or re-submit it if it is still valid and you have rebased it onto ``master`` or merged ``master`` into it.
